src/app/stream-deck/actions.ts

```diff
@@ -0,0 +1,366 @@
+            showSelectionNotification(state, selectionType, () => {
+              dispatch(streamDeckClearSelection());
+            }),
+          500
+        );
+        return;
+      }
+      case 'loadout': {
+        const loadouts = loadoutsSelector(state);
+        const store = storesSelector(state).find((it) => it.id === data.args.character);
+        const loadout = loadouts.find((it) => it.id === data.args.loadout);
+        if (store && loadout) {
+          return dispatch(applyLoadout(store, loadout, { allowUndo: true }));
+        }
+        return;
+      }
+      case 'shareLoadout': {
+        const loadouts = loadoutsSelector(state);
+        const account = currentAccountSelector(state);
+        const accountId = account?.membershipId;
+        const loadout = loadouts.find((it) => it.id === data.args.loadout);
+        if (accountId && loadout) {
+          const shareUrl =
+            loadoutShares.get(loadout) ||
+            (await createLoadoutShare(accountId, convertDimLoadoutToApiLoadout(loadout)));
+          loadoutShares.set(loadout, shareUrl);
+          return dispatch(
+            sendToStreamDeck({
+              shareUrl,
+            })
+          );
+        }
+        return;
+      }
+      case 'freeSlot': {
+        const items = currentStore.items.filter((it) => it.type === data.args.slot);
+        const vaultStore = vaultSelector(state);
+        const pickedItem = items.find((it) => !it.equipped);
+        pickedItem && (await dispatch(moveItemTo(pickedItem, vaultStore!, false)));
+        return;
+      }
+      case 'pullItem': {
+        const allItems = allItemsSelector(state);
+        const vaultStore = vaultSelector(state);
+        const item = allItems.find((it) => it.id === data.args.item);
+        if (item) {
+          if (currentStore.items.includes(item)) {
+            await dispatch(moveItemTo(item, vaultStore!, false));
+          } else {
+            await dispatch(moveItemTo(item, currentStore, false));
+          }
+        }
+      }
+    }
+  };
+}
+
+export const installFarmingObserver = _.once((dispatch) => {
+  observeStore(
+    (state) => state.farming.storeId,
+    (_, newState) => {
+      dispatch(
+        sendToStreamDeck({
+          farmingMode: Boolean(newState),
+        })
+      );
+    }
+  );
+});
+
+// collect and send data to the stream deck
+function refreshStreamDeck(): ThunkResult {
+  return async (dispatch, getState) => {
+    const refreshAction = () => {
+      const state = getState();
+      const store = currentStoreSelector(getState());
+      if (!store) {
+        return;
+      }
+      dispatch(
+        sendToStreamDeck({
+          postmaster: packager.postmaster(store),
+          maxPower: packager.maxPower(store, state),
+          vault: packager.vault(state),
+          metrics: packager.metrics(state),
+        })
+      );
+    };
+    clearInterval(refreshInterval);
+    refreshInterval = setInterval(refreshAction, 30000);
```

Comment:
If you clear the interval every time, it seems like you really want `window.setTimeout`

Reply:
I clear that interval every time the `refreshStreamDeck` that is called so only when the WebSocket is connected, so this is not a `setTimeout` anyways